File: src/DashSpec.Core/Parsing/LayoutParser.cs
using DashSpec.Core.Model;



namespace DashSpec.Core.Parsing;



internal static class LayoutParser

{

    public static LayoutDefinition ParseGrid(TokenReader reader)

    {

        _ = reader.ReadIdent() switch

        {

            "grid" => true,

            _ => throw reader.Unexpected("grid"),

        };



        var props = PropertyBlockParser.Parse(reader, PropertySchemas.LayoutGrid, "layout grid");



        var columns = LayoutDefinition.Default.Columns;

        var gap = LayoutDefinition.Default.GapPx;



        if (props.TryGetValue("columns", out var columnsRaw) &&

            int.TryParse(columnsRaw, out var parsedColumns) &&

            parsedColumns is > 0 and <= 24)

        {

            columns = parsedColumns;

        }



        if (props.TryGetValue("gap", out var gapRaw) &&

            int.TryParse(gapRaw, out var parsedGap) &&

            parsedGap >= 0)

        {

            gap = parsedGap;

        }



        return new LayoutDefinition(columns, gap);

    }



    /// <summary>Tab-level bracket board: layout … end layout.</summary>

    public static LayoutBoardDefinition ParseBoard(TokenReader reader)

    {

        BlockSyntax.BeginBlock(reader);

        reader.SkipNewlines();

        var board = ParseBoardRows(reader, "layout");

        BlockSyntax.ExpectBlockEnd(reader, "layout");

        return board;

    }



    /// <summary>Bracket rows until EOF or end kind (for .dashlayout modules).</summary>

    public static LayoutBoardDefinition ParseBoardRows(TokenReader reader, string? endKind = null, string? endId = null)
    {
        var rows = new List<IReadOnlyList<string>>();
        reader.SkipNewlines();

        while (!reader.IsEof &&
               !reader.IsAt(TokenKind.RBrace) &&
               (endKind is null || !BlockSyntax.IsBlockEnd(reader, endKind, endId)))
        {
            if (!reader.IsAt(TokenKind.LBracket))
            {
                throw reader.Unexpected("[");
            }

            rows.Add(ParseBoardRow(reader));
            reader.SkipNewlines();
        }



        if (rows.Count == 0)

        {

            throw new DashSpecParseException("Layout board requires at least one row [ … ].");

        }



        return new LayoutBoardDefinition(rows);

    }



    public static PlacementDefinition ParsePlacement(TokenReader reader)

    {

        var props = PropertyBlockParser.Parse(reader, PropertySchemas.Placement, "place");



        var row = 1;

        var col = 1;

        var span = 6;



        if (props.TryGetValue("row", out var rowRaw) &&

            int.TryParse(rowRaw, out var parsedRow) &&

            parsedRow > 0)

        {

            row = parsedRow;

        }



        if (props.TryGetValue("col", out var colRaw) &&

            int.TryParse(colRaw, out var parsedCol) &&

            parsedCol > 0)

        {

            col = parsedCol;

        }



        if (props.TryGetValue("span", out var spanRaw))

        {

            span = ParseSpanValue(spanRaw);

        }



        return new PlacementDefinition(row, col, span);

    }



    private static IReadOnlyList<string> ParseBoardRow(TokenReader reader)

    {

        reader.Expect(TokenKind.LBracket);

        reader.SkipNewlines();



        var cells = new List<string>();



        while (!reader.IsAt(TokenKind.RBracket) && !reader.IsEof)

        {

            reader.SkipNewlines();

            if (reader.IsAt(TokenKind.RBracket))

            {

                break;

            }



            cells.Add(reader.ReadIdent());

            reader.SkipNewlines();

        }



        reader.Expect(TokenKind.RBracket);



        if (cells.Count == 0)

        {

            throw new DashSpecParseException("Layout board row [ … ] must list at least one card ref or id.");

        }



        return cells;

    }



    private static int ParseSpanValue(string value) =>

        value.ToLowerInvariant() switch

        {

            "full" => 12,

            "half" => 6,

            "third" => 4,

            _ when int.TryParse(value, out var parsed) && parsed > 0 => parsed,

            _ => 6,

        };

}

File: src/DashSpec.Modeling.Parse/Layout/LayoutParser.fs
namespace DashSpec.Modeling.Parse.Layout

open System
open System.Collections.Generic
open DashSpec.Modeling.Core
open DashSpec.Modeling.Parse
open DashSpec.Modeling.Parse.Lexing

module LayoutParser =

    let private parseBoardRow (reader: TokenReader) =
        reader.Expect TokenKind.LBracket
        reader.SkipNewlines()
        let cells = ResizeArray<string>()
        while not (reader.IsAt TokenKind.RBracket) && not reader.IsEof do
            reader.SkipNewlines()
            if reader.IsAt TokenKind.RBracket then ()
            else
                cells.Add(reader.ReadIdent())
                reader.SkipNewlines()
        reader.Expect TokenKind.RBracket
        if cells.Count = 0 then
            raise (DashSpec.Modeling.Core.DashSpecParseException("Layout board row [ … ] must list at least one card ref or id."))
        cells :> IReadOnlyList<string>

    let private parseGroupBlock (reader: TokenReader) =
        let groupId = reader.ReadIdent()
        if String.IsNullOrWhiteSpace groupId then
            raise (DashSpec.Modeling.Core.DashSpecParseException("Layout group requires an id."))

        BlockSyntax.beginBlock reader
        reader.SkipNewlines()
        let mutable groupTitle: string option = None
        let rows = ResizeArray<IReadOnlyList<string>>()

        while not reader.IsEof && not (BlockSyntax.isBlockEnd reader "group" (Some groupId)) do
            reader.SkipNewlines()
            if BlockSyntax.isBlockEnd reader "group" (Some groupId) then ()
            elif reader.TryKeyword "title" then
                reader.Expect TokenKind.Eq
                groupTitle <- Some(reader.ReadString())
            elif reader.IsAt TokenKind.LBracket then
                rows.Add(parseBoardRow reader)
            else
                raise (reader.Unexpected("[ or title"))

        BlockSyntax.expectBlockEnd reader "group" (Some groupId)

        if rows.Count = 0 then
            raise (DashSpec.Modeling.Core.DashSpecParseException($"Layout group '{groupId}' requires at least one row [ … ]."))

        GroupRow
            { Id = groupId
              Title = groupTitle
              Rows = rows :> IReadOnlyList<_> }

    let private parseBoardEntry (reader: TokenReader) =
        if reader.IsAt TokenKind.LBracket then CardRow(parseBoardRow reader)
        elif reader.TryKeyword "group" then parseGroupBlock reader
        else raise (reader.Unexpected("[ or group"))

    /// Bracket rows and optional groups until EOF or end kind (for .dashlayout modules).
    let parseBoardRows (reader: TokenReader) (endKind: string option) (endId: string option) =
        let entries = ResizeArray<LayoutBoardEntry>()
        reader.SkipNewlines()
        while not reader.IsEof
              && not (reader.IsAt TokenKind.RBrace)
              && (endKind.IsNone || not (BlockSyntax.isBlockEnd reader endKind.Value endId)) do
            entries.Add(parseBoardEntry reader)
            reader.SkipNewlines()
        if entries.Count = 0 then
            raise (DashSpec.Modeling.Core.DashSpecParseException("Layout board requires at least one row [ … ] or group { … }."))
        { Entries = entries :> IReadOnlyList<_>; ModuleScope = None }

    let parseGrid (reader: TokenReader) =
        match reader.ReadIdent() with
        | "grid" -> ()
        | _ -> raise (reader.Unexpected("grid"))

        let props = PropertyBlockParser.parse reader PropertySchemas.layoutGrid "layout grid" false false
        let columns =
            match props.TryGetValue "columns" with
            | true, raw ->
                let mutable parsed = 0
                if System.Int32.TryParse(raw, &parsed) && parsed > 0 && parsed <= 24 then parsed
                else LayoutDefinition.Default.Columns
            | false, _ -> LayoutDefinition.Default.Columns

        let gapPx =
            match props.TryGetValue "gap" with
            | true, raw ->
                let mutable parsed = 0
                if System.Int32.TryParse(raw, &parsed) && parsed >= 0 then parsed
                else LayoutDefinition.Default.GapPx
            | false, _ -> LayoutDefinition.Default.GapPx

        { Columns = columns; GapPx = gapPx }

    let parseBoard (reader: TokenReader) =
        BlockSyntax.beginBlock reader
        reader.SkipNewlines()
        let board = parseBoardRows reader (Some "layout") None
        BlockSyntax.expectBlockEnd reader "layout" None
        board

    let parseSpanValue (value: string) =
        match value.ToLowerInvariant() with
        | "full" -> 12
        | "half" -> 6
        | "third" -> 4
        | _ ->
            let mutable parsed = 0
            if Int32.TryParse(value, &parsed) && parsed > 0 then parsed else 6

    let parsePlacement (reader: TokenReader) =
        let props =
            PropertyBlockParser.parse reader PropertySchemas.placement "place" false false

        let row =
            match props.TryGetValue "row" with
            | true, raw ->
                let mutable parsed = 0
                if Int32.TryParse(raw, &parsed) && parsed > 0 then parsed else 1
            | false, _ -> 1

        let col =
            match props.TryGetValue "col" with
            | true, raw ->
                let mutable parsed = 0
                if Int32.TryParse(raw, &parsed) && parsed > 0 then parsed else 1
            | false, _ -> 1

        let span =
            match props.TryGetValue "span" with
            | true, raw -> parseSpanValue raw
            | false, _ -> 6

        { Row = row; Col = col; Span = span }

    /// Bracket rows until EOF or end kind (string overload for card/toolbar parsers).
    let parseBoardRowsForEndKind (reader: TokenReader) (endKind: string) =
        parseBoardRows reader (Some endKind) None